Panic: File Already Closed Error in Nuclei v3.3.5 with Stats Tag

### Is there an existing issue for this?

- [x] I have searched the existing issues.

### Current Behavior


```console
$ go run -tags=stats . -u honey.scanme.sh -stats 

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v3.3.5

                projectdiscovery.io

[INF] Current nuclei version: v3.3.5 (latest)
[INF] Current nuclei-templates version: v10.0.2 (latest)
[WRN] Scan results upload to cloud is disabled.
[INF] New templates added in latest release: 68
[INF] Templates loaded for current scan: 8655
[INF] Executing 8456 signed templates from projectdiscovery/nuclei-templates
[WRN] Loading 199 unsigned templates for scan. Use with caution.
[INF] Targets loaded for current scan: 1
[INF] Running httpx on input host
[INF] Found 1 URL from httpx
[INF] Templates clustered: 1615 (Reduced 1518 Requests)
panic: write nuclei-stats-20241027002356/events.jsonl: file already closed

goroutine 9445 [running]:
github.com/projectdiscovery/nuclei/v3/pkg/scan/events.(*ScanStatsWorker).AddScanEvent(0x14009c229f0, {{0x0, 0x0}, {0x102960c60, 0x4}, {0x14002a83a90, 0xc}, {0x140000e12c0, 0x42}, 0x1, ...})
        /Users/dogancanbakir/Projects/nuclei/pkg/scan/events/stats_build.go:71 +0x13c
github.com/projectdiscovery/nuclei/v3/pkg/scan/events.AddScanEvent(...)
        /Users/dogancanbakir/Projects/nuclei/pkg/scan/events/stats_build.go:80
github.com/projectdiscovery/nuclei/v3/pkg/tmplexec.(*TemplateExecuter).Execute(0x14003897f00, 0x1401f53f680)
        /Users/dogancanbakir/Projects/nuclei/pkg/tmplexec/exec.go:104 +0x184
github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeAllSelfContained.func1(0x1400786a008)
        /Users/dogancanbakir/Projects/nuclei/pkg/core/executors.go:38 +0x128
created by github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeAllSelfContained in goroutine 1
        /Users/dogancanbakir/Projects/nuclei/pkg/core/executors.go:24 +0x48
exit status 2
```

### Expected Behavior

no panic

### Steps To Reproduce

`go run -tags=stats . -u honey.scanme.sh -stats`

### Relevant log output

_No response_

### Environment

- OS: 
- Nuclei: 
- Go: 


### Anything else?

_No response_

--- a/internal/runner/runner.go
+++ b/internal/runner/runner.go
@@ -392,6 +392,9 @@ func (r *Runner) Close() {
 	if r.tmpDir != "" {
 		_ = os.RemoveAll(r.tmpDir)
 	}
+
+	//this is no-op unless nuclei is built with stats build tag
+	events.Close()
 }
 
 // setupPDCPUpload sets up the PDCP upload writer

--- a/pkg/scan/events/scan_noop.go
+++ b/pkg/scan/events/scan_noop.go
@@ -9,3 +9,6 @@ func AddScanEvent(event ScanEvent) {
 
 func InitWithConfig(config *ScanConfig, statsDirectory string) {
 }
+
+func Close() {
+}

--- a/pkg/scan/events/stats_build.go
+++ b/pkg/scan/events/stats_build.go
@@ -23,6 +23,7 @@ type ScanStatsWorker struct {
 	config    *ScanConfig
 	m         *sync.Mutex
 	directory string
+	file      *os.File
 	enc       *json.Encoder
 }
 
@@ -56,7 +57,7 @@ func (s *ScanStatsWorker) initEventsFile() error {
 	if err != nil {
 		return err
 	}
-	defer f.Close()
+	s.file = f
 	s.enc = json.NewEncoder(f)
 	return nil
 }
@@ -78,4 +79,23 @@ func AddScanEvent(event ScanEvent) {
 		return
 	}
 	defaultWorker.AddScanEvent(event)
+}
+
+// Close closes the file associated with the worker
+func (s *ScanStatsWorker) Close() {
+	s.m.Lock()
+	defer s.m.Unlock()
+
+	if s.file != nil {
+		_ = s.file.Close()
+		s.file = nil
+	}
+}
+
+// Close closes the file associated with the worker
+func Close() {
+	if defaultWorker == nil {
+		return
+	}
+	defaultWorker.Close()
 }

